accepted/2021/statics-in-interfaces/README.md

```diff
@@ -6,8 +6,1470 @@
+          ISubtractionOperators<TimeOnly, TimeOnly, TimeSpan>
+    {
+    }
+
+    public struct TimeSpan
+        : IAdditionOperators<TimeSpan, TimeSpan, TimeSpan>,
+          IAdditiveIdentity<TimeSpan, TimeSpan>,
+          IComparisonOperators<TimeSpan, TimeSpan>,
+          IDivisionOperators<TimeSpan, double, TimeSpan>,
+          IDivisionOperators<TimeSpan, TimeSpan, double>,
+          IMinMaxValue<TimeSpan>,
+          IMultiplicativeIdentity<TimeSpan, double>,
+          IMultiplyOperators<TimeSpan, double, TimeSpan>,
+          IUnaryNegationOperators<TimeSpan, TResult>,
+          ISpanFormattable,
+          ISubtractionOperators<TimeSpan, TimeSpan, TimeSpan>
+    {
+    }
+
+    public struct UInt16
+        : IBinaryInteger<ushort>,
+          IConvertible,
+          IMinMaxValue<ushort>,
+          IUnsignedNumber<ushort>
+    {
+    }
+
+    public struct UInt32
+        : IBinaryInteger<uint>,
+          IConvertible,
+          IMinMaxValue<uint>,
+          IUnsignedNumber<uint>
+    {
+    }
+
+    public struct UInt64
+        : IBinaryInteger<ulong>,
+          IConvertible,
+          IMinMaxValue<ulong>,
+          IUnsignedNumber<ulong>
+    {
+    }
+
+    public struct UIntPtr
+        : IBinaryInteger<nuint>,
+          IMinMaxValue<nuint>,
+          ISerializable,
+          IUnsignedNumber<nuint>
+    {
+    }
+}
+
+namespace System.Numerics
+{
+    public struct Vector<T> : IVector<Vector<T>, T>
+    {
+    }
+
+    public struct Vector2 : IVector<Vector2, float>
+    {
+    }
+
+    public struct Vector3 : IVector<Vector3, float>
+    {
+    }
+
+    public struct Vector4 : IVector<Vector4, float>
+    {
+    }
+}
+
+namespace System.Runtime.Intrinsics
+{
+    public struct Vector64<T> : IVector<Vector64<T>, T>
+    {
+    }
+
+    public struct Vector128<T> : IVector<Vector128<T>, T>
+    {
+    }
+
+    public struct Vector256<T> : IVector<Vector256<T>, T>
+    {
+    }
+}
+```
+
+### Pending Concepts
+
+There are several types which may benefit from some interface support. These include, but aren't limited to:
+* System.Array
+* System.Enum
+* System.Index
+* System.Range
+* System.Tuple
+* System.ValueTuple
+* System.Numerics.BigInteger
+* System.Numerics.Complex
+* System.Numerics.Matrix3x2
+* System.Numerics.Matrix4x4
+* System.Numerics.Plane
+* System.Numerics.Quaternion
+
+Likewise, there are many comments within the defined interfaces above that call out key points that need additional modeling/consideration. Several of these concepts will be determined via user research and outreach while others will be determined via API review and feedback from other area experts. Others may be determined by language or runtime limitations in what is feasible for them to support.
+
+Do we want a way to denote "primitives"? If we have it, should this be language primitives, runtime primitives, or ABI primitives?
+
+Do we want an explicit `IConvertible<TSelf, TOther>` (or `IConvertibleFrom`/`IConvertibleTo`)?
```

Comment:
Guess this comment relates to the comment I gave above. I would think yes.

Reply:
(https://github.com/dotnet/designs/pull/205#discussion_r638217591)
> I have not read everything here but having explicit support for casts/conversion sounds like a better way for this, in many algorithms for image processing you want to move from say byte to short/int and having interfaces for that seems like a good idea.

One issue is that that involves `n` additional interfaces on the primitive types. Each interface implemented by the primitive types will negatively impact startup.
Likewise, there is no way to define say `explicit operator TSelf<TOther>(TOther value) where TOther : INumber<TOther>` and so you can't limit it to say 1 additional interface.

`T.Create(value)` is certainly not as nice as `(T)value`, however it works today and otherwise has all the same "problems" with having to deal with `X.ConvertFrom(y)` vs `y.ConvertTo<X>()`, but with the "added" benefit that you can cover more types of creation (such as exact vs truncating vs saturating, etc).

Reply:
@davidwrighton @jkotas I know optimizations have been done for the interface lookup perf trap. Do we believe there is more we can do there? It seems a shame to compromise design for something that we could potentially optimize during the lifetime of the types.

Reply:
Even if the number of interfaces wasn't a concern, there is still the issue of not being able to define `explicit operator TSelf<TOther>(TOther value) where TOther : INumber<TOther>`. This might just be a language request however.

Reply:
> @davidwrighton @jkotas I know optimizations have been done for the interface lookup perf trap. 

I do not think that this applies here. The generic math feature is using the interfaces as constrains that are expected to be resolved at JIT time. I do not expect there are going to have dynamic casts for generic math in steady state. (Also, we have added cache for the dynamic interface casts that is about as much as we can do for those.)

The way that the generic math feature is designed adds non-trivial static footprint that number of scenarios are going to pay for, even if they do not use it. I am not sure what to additional cost of ConvertFrom/ConvertTo would be, but I suspect it would not be that much given the total bill.

Reply:
This isn't interface lookup, but its more about interface implementation, and the vast number of interfaces involved here. Its just very costly to implement with our current model of how interfaces are loaded at type load time.

There are 3 major costs.
1. Allocation of unique types. Each interface loaded one of these is several hundred bytes.
2. Association of interface types with concrete types like System.Int32. 
3. Sheer volume of methods in System.Private.CoreLib. If we build some sort of IConvertible<TSelf, TOther> interface and make all the simple numeric (signed and unsigned integer types + float+ double) types implement the full cross product of them, we're talking 10^2 different interfaces, and the same number of additional methods. Each method is something like 100 bytes in the file, so we're looking at a somewhat unfortunate combinatorial expansion problem.

Reply:
One concern I have with using the `Create<TOther>` pattern over dedicated conversion interfaces is that it is not extensible, and misuse is a runtime failure since [the actual support for `TOther` depends on the implementation](https://github.com/dotnet/runtime/blob/c0662e8129beaf93b8050d39a863cc6d16a0308c/src/libraries/System.Private.CoreLib/src/System/Int32.cs#L488).

For example, consider the following function:

```csharp
static TTo Convert<TFrom, TTo>(TFrom from)
    where TTo : INumber<TTo>
    where TFrom : INumber<TFrom>
    => TTo.Create(from);
```

This works fine for types explicitly supported by `TTo`, IE: `Convert<short, int>(x)`

But if I define my own `INumber` implementation, say an `Int128` type and try `Convert<Int128, int>(x)` it will compile just fine but will fail at runtime with `NotSupportedException`.

Maybe `Create<TOther>` should be restricted to the primitive numeric types?